hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomFeeExemptions.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.transfer.customfees;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.transaction.CustomFee;
+
+/**
+ * Our policy for payer exemptions is the following:
+ *
+ * <ul>
+ *   <li>A token's treasury is exempt from all its custom fees.
+ *   <li>A fee collection account is exempt from any fee for which it would be the collector.
+ *   <li>A fee collection account is exempt from any of its token's fees with {@code
+ *       all_collectors_are_exempt=true}.
+ * </ul>
+ */
+public class CustomFeeExemptions {
+    private CustomFeeExemptions() {
+        throw new UnsupportedOperationException("Utility Class");
+    }
+    /**
+     * Given the fee metadata for a token, and one of this token's custom fees, returns whether the
+     * given payer is exempt from the specific custom fee provided.
+     * Payer is exempt if:
+     * <ul>
+     *     <li>the payer is the treasury of the token
+     *     <li>the payer is the fee collector of the fee
+     *     <li>if allCollectorsAreExempt set to true and payer is collector for any fee on token
+     * </ul>
+     *
+     * @param feeMeta metadata for the token that "owns" the specific custom fee
+     * @param fee the fee to check for a payer exemption
+     * @param sender the potential fee payer
+     * @return whether the payer is exempt from the fee
+     */
+    public static boolean isPayerExempt(final CustomFeeMeta feeMeta, final CustomFee fee, final AccountID sender) {
+        if (feeMeta.treasuryId().equals(sender)) {
+            return true;
+        }
+        if (fee.feeCollectorAccountIdOrElse(AccountID.DEFAULT).equals(sender)) {
+            return true;
+        }
+        if (fee.allCollectorsAreExempt()) {
+            return isPayerCollectorFor(feeMeta, sender);
+        } else {
+            // If payer isn't the treasury or the collector of a fee without
+            // a global collector exemption, then it must pay, nothing more to check
+            return false;
+        }
```

Comment:
No need for `else` part here, if I'm reading right. Not a big deal, but maybe less braces to read:

```java
```
```suggestion
        if (fee.allCollectorsAreExempt()) {
            return isPayerCollectorFor(feeMeta, sender);
        } 
        
        // If payer isn't the treasury or the collector of a fee without
        // a global collector exemption, then it must pay, nothing more to check
        return false;
```

Reply:
Changed this logic